### PR TITLE
Add bootstrap and placebo to synthdid_se

### PR DESCRIPTION
**Pull request by erikcs**
Update: `synthdid_se(estimate, method = c("bootstrap", "jackknife", "placebo"))` ( can also be accessed with `vcov`).

The default standard error is changed from jackknife to bootstrap.

### CODE REVIEW

**Comment by davidahirshberg** [comment]
1. Could you add some tests demonstrating coverage?
2. What is the 'estimator' attribute for?

File: R/synthdid.R [hunk 1]
```diff
@@ -179,21 +184,67 @@ synthdid_effect_curve = function(estimate) {
...
+    if (is.null(treated.fraction)) { treated.fraction = 1 - setup$T0 / ncol(setup$Y) }
+    n1 = floor(nrow(setup$Y) * treated.fraction)
+    setup.bs = setup
+    theta = function(ind) {
+      Y_b = rbind(setup$Y[-ind,], setup$Y[ind,])
+      # X_b = rbind(setup$X[-ind, ,], setup$X[ind, ,])
+      n_0 = nrow(setup$Y) - n1
+      setup.bs$Y = Y_b
+      setup.bs$N0 = n_0
+      # setup.bs$X = X_b
```
Comment (davidahirshberg) [inline R/synthdid.R:245]: Why are covariates commented out for the bootstrap?

Comment (erikcs) [reply R/synthdid.R:245]: Pending an rbind-workaround for a 3d array, suggestion for that?

Comment (davidahirshberg) [reply R/synthdid.R:245]: cat

File: R/utils.R [hunk 2]
```diff
@@ -14,6 +14,25 @@ jackknife = function(x, theta) {
...
+  for (i in 1:B) {
+    ind = sample(1:n, k, replace = replace)
+    u[i] = theta(ind)
+  }
+  boot.se = sqrt(mean((c(estimate) - u)^2, na.rm = TRUE))
+  # boot.se=sqrt(var(u, na.rm = TRUE))
+
+  boot.se
+}
+
```
Comment (davidahirshberg) [inline R/utils.R:35]: This is a pretty confusing interface. Maybe replace it with two functions.
1. bootstrap(x,theta,replications) where x and theta match what is used for the jackknife and we always sample a list of the same length as x with replacement
2. subsample(x,theta,size,replications) where we subsample without replacement

File: R/synthdid.R (hunk 1 above)
Comment (davidahirshberg) [inline R/synthdid.R:245]: Shouldn't we be working with just the control units here, and subsampling N_1 of the control units (without replacement) to work with as if they were treated?  What is with the treated.fraction stuff?

Comment (erikcs) [reply R/synthdid.R:245]: I did not understand the placebo-SE from code nor paper; if you add a R skeleton here I can incorporate that

**Review by davidahirshberg — COMMENTED** [review]
Before going live with this stuff, we should have some tests showing that these estimates are similar (the placebo only with homoskedastic noise) and that they get near-nominal coverage.

Re organization, we should probably implement our 3 se estimators as 3 different functions, 
letting synthdid_se refer to the bootstrap one for backward compatibility, and implementing the generic vcov in terms of the bootstrap as well. Docstrings for each should mention the other options and refer to the appopriate 'Algorithm' box in the paper; the docstring for the jackknife should warn against using it for SC with reference to paper Section 5.

Let's take B=200 replications for the bootstrap by default.

File: R/synthdid.R (hunk 1 above)
Comment (davidahirshberg) [reply R/synthdid.R:245]: placebo.estimates = rep(NA, B)
for(b in 1:B) {
  Y = setup$Y[sample(1:setup$N0), ]
  N1 = nrow(setup$Y)-setup$N0
  placebo.estimates[b] = synthdid_estimate(Y,N0=nrow(Y)-N1,  setup$T0)
}
return sd(placebo.estimates)

subsample(1:N0, theta, sizereplications=B)
   where
theta = function(ind) {

**Comment by erikcs** [comment]
> Could you add some tests demonstrating coverage?

@ 025291f

```R
library(synthdid)
estimators = list("sc_estimate", "did_estimate", "synthdid_estimate")
CI.methods = c("jackknife", "bootstrap", "placebo")
res = 0
for (estimator in estimators) {
  for (CI.method in CI.methods) {
    cov = replicate(100, {
      setup = synthdid:::random.low.rank()
      estimate = do.call(estimator, setup[-2])
      se = synthdid_se(estimate, method = CI.method, repetitions = 200)
      1 > estimate - 1.96*se && 1 < estimate + 1.96*se
    })
    res = c(res, c(paste0(estimator, CI.method, ": ", mean(cov))))
  }
}
matrix(res[-1], length(res[-1]), 1)
      [,1]                              
 [1,] "sc_estimatejackknife: 0.96"      
 [2,] "sc_estimatebootstrap: 0.76"      
 [3,] "sc_estimateplacebo: 0.17"        
 [4,] "did_estimatejackknife: 0.75"     
 [5,] "did_estimatebootstrap: 0.72"     
 [6,] "did_estimateplacebo: 0.67"       
 [7,] "synthdid_estimatejackknife: 0.92"
 [8,] "synthdid_estimatebootstrap: 0.92"
 [9,] "synthdid_estimateplacebo: 0.92"
```

File: R/utils.R (hunk 2 above)
Comment (davidahirshberg) [reply R/utils.R:36]: Sorry, I should have given you the implementations. These functions are essentially standard and encode invariants useful when thinking about the method. Please use the following.

bootstrap = function(x,theta,replications) {
    replicate(replications, theta(sample(x, replace=TRUE)))
}
subsample = function(x,theta,size,replications) {
   replicate(replications, theta(sample(x, size, replace=FALSE)))
}

But we're not actually subsampling, so we don't need the second. Sorry about that.
In terms of these, we can implement standard errors as follows:

bootstrap_se = function(estimate, replications=200) { 
   theta = function(ind) { synthdid_estimate(Y=setup$Y[sort(ind)], N0=sum(ind <= setup$N0), T0=setup$T0, X=setup$X[sort(ind), ,], <options> }
sd(bootstrap(1:nrow(Y),  theta, replications=replications))
}

placebo_se = function(estimate) {
  N1=nrow(setup$Y)-setup$N0
  theta = function(ind) { synthdid_estimate(Y=setup$Y[ind], N0=length(ind)-N1,  T0=setup$T0,  X=setup$X[ind, ,], <options> }
  sd(replicate(replications, theta(sample(1:setup$N0))))
}

Comment (erikcs) [reply R/utils.R]: Ok, it is updated

Comment (davidahirshberg) [reply R/utils.R]: The <options> should be the same ones passed to produce the original estimate, so the optimizer gets the same stopping parameters, we use sc or did if that is what we used in the first place, etc. 

And we should return NA when there is one treated unit for the bootstrap and jackknife, but not the placebo --- that's what the placebo estimate is for.

**Comment by davidahirshberg** [comment]
Docs for synthdid_se are out of date. Could you update them with the info that is now in the vcov docs?
How is the coverage now?

**Comment by erikcs** [comment]
> Docs for synthdid_se are out of date. Could you update them with the info that is now in the vcov docs?
> How is the coverage now?

```R
library(synthdid)
estimators = list("sc_estimate", "did_estimate", "synthdid_estimate")
CI.methods = c("jackknife", "bootstrap", "placebo")
res = 0
for (estimator in estimators) {
  for (CI.method in CI.methods) {
    cat(estimator, CI.method,"\n")
    cov = replicate(100, {
      setup = synthdid:::random.low.rank()
      estimate = do.call(estimator, setup[-2])
      se = synthdid_se(estimate, method = CI.method)
      1 > estimate - 1.96*se && 1 < estimate + 1.96*se
    })
    res = c(res, c(paste0(estimator, CI.method, ": ", mean(cov))))
  }
}
matrix(res[-1], length(res[-1]), 1)
      [,1]                              
 [1,] "sc_estimatejackknife: 0.94"      
 [2,] "sc_estimatebootstrap: 0.7"       
 [3,] "sc_estimateplacebo: 0.17"        
 [4,] "did_estimatejackknife: 0.79"     
 [5,] "did_estimatebootstrap: 0.77"     
 [6,] "did_estimateplacebo: 0.77"       
 [7,] "synthdid_estimatejackknife: 0.92"
 [8,] "synthdid_estimatebootstrap: 0.97"
 [9,] "synthdid_estimateplacebo: 0.97"
```

**Comment by davidahirshberg** [comment]
We'll need to look into what is going on re: coverage for DID and SC. 
--Are they calling the right functions with the right options?
--Is it a bias issue? Try doing the same with a 2-way-fixed effects model. 
100 replications is enough to detect severe coverage problems, but not enough to see if we're getting nominal coverage. 
We should increase this --- 1000 should more or less do it.

**Comment by erikcs** [comment]
> We'll need to look into what is going on re: coverage for DID and SC.
> --Are they calling the right functions with the right options?

I don't see how that's not the case?

> --Is it a bias issue? Try doing the same with a 2-way-fixed effects model.

You have a DGP suggestion?

> 100 replications is enough to detect severe coverage problems, but not enough to see if we're getting nominal coverage.
> We should increase this --- 1000 should more or less do it.

1000:
```
 [1,] "sc_estimatejackknife: 0.97"       
 [2,] "sc_estimatebootstrap: 0.733"      
 [3,] "sc_estimateplacebo: 0.193"        
 [4,] "did_estimatejackknife: 0.827"     
 [5,] "did_estimatebootstrap: 0.785"     
 [6,] "did_estimateplacebo: 0.75"        
 [7,] "synthdid_estimatejackknife: 0.929"
 [8,] "synthdid_estimatebootstrap: 0.942"
 [9,] "synthdid_estimateplacebo: 0.942"  
```

**Comment by erikcs** [comment]
With

```R
random.low.rank = function() {
  n_0 <- 100
  n_1 <- 10
  T_0 <- 120
  T_1 <- 20
  n <- n_0 + n_1
  T <- T_0 + T_1
  tau <- 1
  sigma <- .5
  rank <- 2
  rho <- 0.7
  var <- outer(1:T, 1:T, FUN=function(x, y) rho^(abs(x-y)))
  
  W <- (1:n > n_0) %*% t(1:T > T_0)
  U <- matrix(rpois(rank * n, sqrt(1:n) / sqrt(n)), n, rank)
  V <- matrix(rpois(rank * T, sqrt(1:T) / sqrt(T)), T, rank)
  alpha <- outer(10*(1:n)/n, rep(1,T))
  beta <-  outer(rep(1,n), 10*(1:T)/T)
  # mu <- U %*% t(V) + alpha + beta
  mu = 0
  # error <- mvtnorm::rmvnorm(n, sigma = var, method = "chol")
  error = matrix(rnorm(110*140),110,140)
  Y <- mu + tau * W  + sigma * error
  rownames(Y) = 1:n
  colnames(Y) = 1:T
  list(Y=Y, L=mu, N0=n_0, T0=T_0)
}
```
250 reps:  

```
 [1,] "sc_estimatejackknife: 0.948"      
 [2,] "sc_estimatebootstrap: 0.94"       
 [3,] "sc_estimateplacebo: 0.952"        
 [4,] "did_estimatejackknife: 0.936"     
 [5,] "did_estimatebootstrap: 0.944"     
 [6,] "did_estimateplacebo: 0.956"       
 [7,] "synthdid_estimatejackknife: 0.936"
 [8,] "synthdid_estimatebootstrap: 0.976"
 [9,] "synthdid_estimateplacebo: 0.936" 
```

**Comment by davidahirshberg** [comment]
> You have a DGP suggestion?
Y_it = alpha_i + beta_t + epsilon_it, epsilon_it ~ N(0,1)

**Comment by erikcs** [comment]
> > You have a DGP suggestion?
> > Y_it = alpha_i + beta_t + epsilon_it, epsilon_it ~ N(0,1)

```
library(synthdid)
random.low.rank = function() {
  n_0 <- 100
  n_1 <- 10
  T_0 <- 120
  T_1 <- 20
  n <- n_0 + n_1
  T <- T_0 + T_1
  tau <- 1
  sigma <- .5
  rank <- 2
  rho <- 0.7
  var <- outer(1:T, 1:T, FUN=function(x, y) rho^(abs(x-y)))

  W <- (1:n > n_0) %*% t(1:T > T_0)
  # U <- matrix(rpois(rank * n, sqrt(1:n) / sqrt(n)), n, rank)
  # V <- matrix(rpois(rank * T, sqrt(1:T) / sqrt(T)), T, rank)
  alpha <- outer(10*(1:n)/n, rep(1,T))
  beta <-  outer(rep(1,n), 10*(1:T)/T)
  # mu <- U %*% t(V) + alpha + beta
  mu = 0
  # error <- mvtnorm::rmvnorm(n, sigma = var, method = "chol")
  error = matrix(rnorm(110*140),110,140)
  # Y <- mu + tau * W  + sigma * error
  Y <- alpha + beta + error
  rownames(Y) = 1:n
  colnames(Y) = 1:T
  list(Y=Y, L=mu, N0=n_0, T0=T_0)
}
estimators = list("sc_estimate", "did_estimate", "synthdid_estimate")
CI.methods = c("jackknife", "bootstrap", "placebo")
res = 0
for (estimator in estimators) {
  for (CI.method in CI.methods) {
    cov = replicate(1000, {
      setup = random.low.rank()
      estimate = do.call(estimator, setup[-2])
      se = synthdid_se(estimate, method = CI.method)
      0 > estimate - 1.96*se && 0 < estimate + 1.96*se
    })
    res = c(res, c(paste0(estimator, CI.method, ": ", mean(cov))))
  }
}
matrix(res[-1], length(res[-1]), 1)
 [1,] "sc_estimatejackknife: 0.001"      
 [2,] "sc_estimatebootstrap: 0"          
 [3,] "sc_estimateplacebo: 0"            
 [4,] "did_estimatejackknife: 0.922"     
 [5,] "did_estimatebootstrap: 0.932"     
 [6,] "did_estimateplacebo: 0.952"       
 [7,] "synthdid_estimatejackknife: 0.919"
 [8,] "synthdid_estimatebootstrap: 0.954"
 [9,] "synthdid_estimateplacebo: 0.96" 
```

**Comment by erikcs** [comment]
```
library(synthdid)
random.low.rank = function() {
  n_0 <- 100
  n_1 <- 10
  T_0 <- 120
  T_1 <- 20
  n <- n_0 + n_1
  T <- T_0 + T_1
  tau <- 1
  sigma <- .5
  rank <- 2
  rho <- 0.7
  var <- outer(1:T, 1:T, FUN=function(x, y) rho^(abs(x-y)))

  W <- (1:n > n_0) %*% t(1:T > T_0)
  # U <- matrix(rpois(rank * n, sqrt(1:n) / sqrt(n)), n, rank)
  # V <- matrix(rpois(rank * T, sqrt(1:T) / sqrt(T)), T, rank)
  alpha <- outer(10*(1:n)/n, rep(1,T))
  alpha = alpha[sample(1:nrow(alpha)), ]
  beta <-  outer(rep(1,n), 10*(1:T)/T)
  # mu <- U %*% t(V) + alpha + beta
  mu = 0
  # error <- mvtnorm::rmvnorm(n, sigma = var, method = "chol")
  error = matrix(rnorm(110*140),110,140)
  # Y <- mu + tau * W  + sigma * error
  Y <- alpha + beta + error
  rownames(Y) = 1:n
  colnames(Y) = 1:T
  list(Y=Y, L=mu, N0=n_0, T0=T_0)
}
estimators = list("sc_estimate", "did_estimate", "synthdid_estimate")
CI.methods = c("jackknife", "bootstrap", "placebo")
res = 0
for (estimator in estimators) {
  for (CI.method in CI.methods) {
    cov = replicate(1000, {
      setup = random.low.rank()
      estimate = do.call(estimator, setup[-2])
      se = synthdid_se(estimate, method = CI.method)
      0 > estimate - 1.96*se && 0 < estimate + 1.96*se
    })
    res = c(res, c(paste0(estimator, CI.method, ": ", mean(cov))))
  }
}
matrix(res[-1], length(res[-1]), 1)
      [,1]                               
 [1,] "sc_estimatejackknife: 1"          
 [2,] "sc_estimatebootstrap: 0.919"      
 [3,] "sc_estimateplacebo: 0.95"         
 [4,] "did_estimatejackknife: 0.933"     
 [5,] "did_estimatebootstrap: 0.912"     
 [6,] "did_estimateplacebo: 0.953"       
 [7,] "synthdid_estimatejackknife: 0.927"
 [8,] "synthdid_estimatebootstrap: 0.98" 
 [9,] "synthdid_estimateplacebo: 0.949"  
```

**Comment by erikcs** [comment]
With the latest low.rank dgp:

```
set.seed(42)
library(synthdid)

estimators = list("sc_estimate", "did_estimate", "synthdid_estimate")
CI.methods = c("jackknife", "bootstrap", "placebo")
res = 0
for (estimator in estimators) {
  for (CI.method in CI.methods) {
    cov = replicate(1000, {
      setup = synthdid:::random.low.rank()
      estimate = do.call(estimator, setup[-2])
      se = synthdid_se(estimate, method = CI.method)
      1 > estimate - 1.96*se && 1 < estimate + 1.96*se
    })
    res = c(res, c(paste0(estimator, CI.method, ": ", mean(cov))))
  }
}
matrix(res[-1], length(res[-1]), 1)
      [,1]                               
 [1,] "sc_estimatejackknife: 1"          
 [2,] "sc_estimatebootstrap: 0.954"      
 [3,] "sc_estimateplacebo: 0.939"        
 [4,] "did_estimatejackknife: 0.933"     
 [5,] "did_estimatebootstrap: 0.92"      
 [6,] "did_estimateplacebo: 0.951"       
 [7,] "synthdid_estimatejackknife: 0.932"
 [8,] "synthdid_estimatebootstrap: 0.938"
 [9,] "synthdid_estimateplacebo: 0.937" 
```

**Comment by davidahirshberg** [comment]
Looks good.

**Review by davidahirshberg — APPROVED** [review]

**Pull request merged**